keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1615 @@
+
+- nftables mode has unit test parity with iptables
+
+- An nftables-mode e2e job exists, and passes
+
+#### Beta
+
+- At least two releases since Alpha.
+
+- The nftables mode has seen real-world usage, no major bugs.
+
+- A "kube-proxy mode-switching" e2e job exists, to confirm that you
+  can redeploy kube-proxy in a different mode in an existing cluster.
+  (eg, it tests "`kube-proxy --cleanup`")
+
+- An nftables e2e periodic perf/scale job exists, and shows
+  performance as good as iptables and ipvs.
+
+- No UNRESOLVED sections in the KEP. (In particular, we have figured
+  out what sort of "API" we will offer for integrating third-party
+  nftables rules.)
+
+#### GA
+
+- At least two releases since Beta.
+
+- The nftables mode has no bugs / regressions that would make us
+  hesitate to recommend it.
+
+- We have at least the start of a plan for the next steps (changing
+  the default mode, deprecating the old backends, etc).
+
+### Upgrade / Downgrade Strategy
+
+The new mode should not introduce any upgrade/downgrade problems,
+excepting that you can't downgrade or feature-disable a cluster using
+the new kube-proxy mode without switching it back to `iptables` or
+`ipvs` first.
+
+When rolling out or rolling back the feature, it should be safe to
+enable the feature gate and change the configuration at the same time,
+since nothing cares about the feature gate except for kube-proxy
+itself. Likewise, it is expected to be safe to roll out the feature in
+a live cluster, even though this will result in different proxy modes
+running on different nodes, because Kubernetes service proxying is
+defined in such a way that no node needs to be aware of the
+implementation details of the service proxy implementation on any
+other node.
+
+(However, see the notes below in [Feature Enablement and
+Rollback](#feature-enablement-and-rollback) about stale rule cleanup
+when switching modes.)
+
+### Version Skew Strategy
+
+The feature is isolated to kube-proxy and does not introduce any API
+changes, so the versions of other components do not matter.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+The administrator must enable the feature gate to make the feature
+available, and then must run kube-proxy with the
+`--proxy-mode=nftables` flag.
+
+Kube-proxy does not delete its rules on exit (to avoid service
+interruptions when restarting/upgrading kube-proxy, or if it crashes).
+This means that when switching between proxy modes, it is necessary
+for the administrator to ensure that the rules created by the old
+proxy mode get deleted. (Failure to do so may result in stale service
+rules being left behind for an arbitrarily long time.) The simplest
+way to do this is to reboot each node when switching from one proxy
+mode to another, but it is also possible to run kube-proxy in "cleanup
+and exit" mode, eg:
```

Comment:
This seems rather cumbersome? Can we somehow detect that proxy mode has changed and help user with the transition?

Reply:
kube-proxy used to do that and we decided it was terrible and removed the code...
I'll have to dig up the old discussions and figure out if those reasons still apply.

Reply:
oh, look, we wrote a KEP: [KEP-2448 Remove kube-proxy's automatic clean up logic](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2448-Remove-kube-proxy-automatic-clean-up-logic).

So the problem there was that the userspace, iptables, and ipvs proxy modes used some of the same chain names (especially, they all use `KUBE-SERVICES`), so trying to clean up the non-active modes would result in deleting some of the chains used by the active mode too, causing disruption until they were recreated.

This would not apply when switching to a backend based on an entirely different technology though, so we _could_ automatically clean up the iptables rules when switching to nftables mode, and vice versa.